File: example/unity_client/Assets/Scripts/AStarPathFinder.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class AStarPathFinder : IEnumerable
{
	public class Node
	{
		public Node(AStarPathFinder pathFinder, Node parent, Vector2Int from, int pathCost)
		{
			path_finder = pathFinder;
			position = from;
			expect_cost += Mathf.Abs(pathFinder.destination.x - from.x);
			expect_cost += Mathf.Abs(pathFinder.destination.y - from.y);
			path_cost = pathCost;
			index = from.y * path_finder.dungeon.width + from.x;
			this.parent = parent;
		}

		public void FindPath()
		{
			if (true == path_finder.open_nodes.ContainsKey(index))
			{
				path_finder.open_nodes.Remove(index);
			}

			if (true == path_finder.close_nodes.ContainsKey(index))
			{
				//Debug.Log("can't find path(position:" + position.ToString() + ")");
				return;
			}

			path_finder.close_nodes.Add(index, this);
			if (position == path_finder.destination)
			{
				Node trace = this;
				while (null != trace.parent)
				{
					path_finder.path.Insert(0, trace.position);
					trace = trace.parent;
				}
				path_finder.path.Insert(0, trace.position);
				return;
			}

			List<Node> children = new List<Node>();
			foreach (Vector2Int offset in LOOKUP_OFFSETS)
			{
				Vector2Int childPosition = new Vector2Int(position.x + offset.x, position.y + offset.y);
				if (0 > childPosition.x || 0 > childPosition.y || path_finder.dungeon.width <= childPosition.x || path_finder.dungeon.height <= childPosition.y)
				{
					continue;
				}

				Tile.Data tile = path_finder.dungeon.GetTileData(childPosition.x, childPosition.y);
				if (Message.DungeonTileType.Wall == tile.type)
				{
					//Debug.Log("not tile(child position:x" + childPosition.x + ", y:" + childPosition.y +")");
					continue;
				}

				Node childNode = new Node(path_finder, this, childPosition, path_cost + 1);
				if (true == path_finder.close_nodes.ContainsKey(childNode.index))
				{
					//Debug.Log ("already close node(child position:x" + childPosition.x + ", y:" + childPosition.y +")");
					continue;
				}
				if (true == path_finder.open_nodes.ContainsKey(childNode.index))
				{
					Node openNode = path_finder.open_nodes[childNode.index];
					if (openNode.path_cost < this.path_cost)
					{
						this.path_cost = openNode.path_cost + 1;
						this.parent = openNode;
						continue;
					}
				}
				else
				{
					path_finder.open_nodes.Add(childNode.index, childNode);
				}
				children.Add(childNode);
			}
			children.Sort(delegate (Node lhs, Node rhs) {
				return lhs.cost.CompareTo(rhs.cost);
			});
			foreach (Node child in children)
			{
				child.FindPath();
				if (0 < path_finder.path.Count)
				{
					return;
				}
			}
			//Debug.Log ("cant find destination(x:" + position.x + ", y:" + position.y + ")");
		}

		public int index = 0;
		public int path_cost = 0;
		public int expect_cost = 0;
		public int cost
		{
			get { return path_cost + expect_cost; }
		}
		public Node parent = null;
		public Vector2Int position;
		private AStarPathFinder path_finder = null;
		
		static Vector2Int[] LOOKUP_OFFSETS = {
			new Vector2Int(-1,  0),
			new Vector2Int( 1,  0),
			new Vector2Int( 0, -1),
			new Vector2Int( 0,  1)
		};
	}

	public Component.Dungeon dungeon;
	public Dictionary<int, Node> open_nodes = new Dictionary<int, Node>();
	public Dictionary<int, Node> close_nodes = new Dictionary<int, Node>();
	public Vector2Int destination = Vector2Int.zero;
	public List<Vector2Int> path = new List<Vector2Int>();

	public AStarPathFinder(Component.Dungeon dungeon, Vector2Int from, Vector2Int to)
	{
		this.dungeon = dungeon;
		destination = to;
		Node node = new Node(this, null, from, 0);
		node.FindPath();		
	}

	public IEnumerator GetEnumerator()
	{
		if (0 == path.Count)
		{
			yield break;
		}
		Vector2Int position = path[0];
		path.RemoveAt(0);
		yield return position;
	}
}

File: example/LobbyServer/Component/Dungeon/AStarPathFinder.cpp
#include "AStarPathFinder.h"
#include "Dungeon.h"
#include <algorithm>

const Vector2Int AStarPathFinder::Node::LOOKUP_OFFSETS[4] =
{
	Vector2Int(-1,  0),
	Vector2Int(1,  0),
	Vector2Int(0, -1),
	Vector2Int(0,  1)
};

AStarPathFinder::Node::Node(AStarPathFinder* pathFinder, std::shared_ptr<Node> parent, const Vector2Int& from, int pathCost)
	: path_finder(pathFinder)
	, parent(parent)
	, position(from)
	, path_cost(pathCost)
	, expect_cost(0)
	, index(0)
{
	index = from.y * path_finder->dungeon->rect.width + from.x;
	expect_cost += std::abs(path_finder->destination.x - from.x);
	expect_cost += std::abs(path_finder->destination.y - from.y);
}

void AStarPathFinder::Node::FindPath()
{
	{
		auto itr = path_finder->open_nodes.find(index);
		if (path_finder->open_nodes.end() != itr)
		{
			path_finder->open_nodes.erase(itr);
		}
	}

	{
		auto itr = path_finder->close_nodes.find(index);
		if (path_finder->close_nodes.end() != itr)
		{
			return;
		}
	}

	path_finder->close_nodes.insert(std::make_pair(index, shared_from_this()));
	if (position == path_finder->destination)
	{
		std::shared_ptr<Node> trace = shared_from_this();
		while (nullptr != trace->parent)
		{
			path_finder->path.push_front(trace->position);
			trace = trace->parent;
		}
		path_finder->path.push_front(trace->position);
		return;
	}

	std::list<std::shared_ptr<Node>> children;
	for (const Vector2Int& offset : LOOKUP_OFFSETS)
	{
		Vector2Int childPosition(position.x + offset.x, position.y + offset.y);
		std::shared_ptr<Tile> tile = path_finder->dungeon->GetTile(childPosition.x, childPosition.y);
		if(nullptr == tile)
		{
			continue;
		}
		
		if (Tile::Type::Wall == tile->type)
		{
			continue;
		}

		std::shared_ptr<Node> childNode = std::make_shared<Node>(path_finder, shared_from_this(), childPosition, path_cost + 1); 
		{
			auto itr = path_finder->close_nodes.find(childNode->index);
			if (path_finder->close_nodes.end() != itr)
			{
				//Debug.Log ("already close node(child position:x" + childPosition.x + ", y:" + childPosition.y +")");
				continue;
			}
		}
		{
			auto itr = path_finder->open_nodes.find(childNode->index);
			if (path_finder->open_nodes.end() != itr)
			{
				std::shared_ptr<Node> openNode = itr->second;
				if (openNode->path_cost < this->path_cost)
				{
					this->path_cost = openNode->path_cost + 1;
					this->parent = openNode;
					continue;
				}
			}
			else
			{
				path_finder->open_nodes.insert(std::make_pair(childNode->index, childNode));
			}
		}
		children.push_back(childNode);
	}
	children.sort([](const std::shared_ptr<Node>& lhs, const std::shared_ptr<Node>& rhs)
	{
		return lhs->GetCost() < rhs->GetCost();
	});

	for(std::shared_ptr<Node>& child : children)
	{
		child->FindPath();
		if (0 < path_finder->path.size())
		{
			return;
		}
	}
	//Debug.Log ("cant find destination(x:" + position.x + ", y:" + position.y + ")");
}

int AStarPathFinder::Node::GetCost() const
{
	return path_cost + expect_cost;
}


AStarPathFinder::AStarPathFinder(std::shared_ptr<Component::Dungeon::Data>& dungeon, const Vector2Int& from, const Vector2Int& to)
	: dungeon(dungeon)
	, destination(to)
{
	std::shared_ptr<Node> node = std::make_shared<Node>(this, nullptr, from, 0);
	node->FindPath();
}